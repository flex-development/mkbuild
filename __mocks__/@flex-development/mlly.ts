/**
 * @file Mocks - @flex-development/mlly
 * @module mocks/flex-development/mlly
 * @see https://github.com/flex-development/mlly
 */

/**
 * [`@flex-development/mlly`][1] module type.
 *
 * [1]: https://github.com/flex-development/mlly
 */
type Actual = typeof import('@flex-development/mlly')

/**
 * `@flex-development/mlly` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/mlly')

export const CONDITIONS = actual.CONDITIONS
export const RESOLVE_EXTENSIONS = actual.RESOLVE_EXTENSIONS
export const fillModules = vi.fn(actual.fillModules)
export const findStaticImports = vi.fn(actual.findStaticImports)
export const getSource = vi.fn(actual.getSource)
export const readPackageJson = vi.fn(actual.readPackageJson)
export const resolveModule = vi.fn(actual.resolveModule)
export const resolveModules = vi.fn(actual.resolveModules)
export const toDataURL = vi.fn(actual.toDataURL)
export const toURL = vi.fn(actual.toURL)
