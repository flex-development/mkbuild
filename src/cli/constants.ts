/**
 * @file CLI - Constants
 * @module mkbuild/cli/constants
 */

import pkg from '#pkg' assert { type: 'json' }

/**
 * Array containing choices for boolean flags.
 *
 * @internal
 *
 * @const {string[]} CHOICES_BOOLEAN
 */
export const CHOICES_BOOLEAN: string[] = ['false', 'true', '0', '1']

/**
 * CLI application name.
 *
 * @internal
 *
 * @const {string} CLI_NAME
 */
export const CLI_NAME: string = pkg.name.replace(/.*\//, '')
