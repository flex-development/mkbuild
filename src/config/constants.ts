/**
 * @file Config - Constants
 * @module mkbuild/config/constants
 */

/**
 * Default ignore patterns.
 *
 * @see https://github.com/mrmlnc/fast-glob#ignore
 *
 * @const {string[]} IGNORE_PATTERNS
 */
export const IGNORE_PATTERNS: string[] = [
  '**/.DS_Store',
  '**/.npmignore',
  '**/.yarnignore',
  '**/**.mdx',
  '**/**.stories.{j,t}sx',
  '**/{__mocks__,__snapshots__,__tests__}/**'
]

/**
 * Module extensions checked when attempting to resolve path aliases and module
 * specifiers.
 *
 * @const {string[]} MODULE_EXTENSIONS
 */
export const MODULE_EXTENSIONS: string[] = [
  '.cjs',
  '.cts',
  '.js',
  '.json',
  '.jsx',
  '.mdx',
  '.mjs',
  '.mts',
  '.ts',
  '.tsx'
]

/**
 * `require` and `require.resolve` statement regex.
 *
 * @todo prevent matching non-statements (e.g. comments, arbitrary strings)
 *
 * @const {RegExp} REQUIRE_REGEX
 */
export const REQUIRE_REGEX: RegExp =
  /(?:require\(|require\.resolve\()["']([^"']*)["']\)/g
