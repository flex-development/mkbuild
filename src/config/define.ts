/**
 * @file Config - defineBuildConfig
 * @module mkbuild/config/defineBuildConfig
 */

import type { Config } from '#src/interfaces'

/**
 * Helper for defining a build configuration object.
 *
 * @param {Config} [config={}] - Build configuration options
 * @return {Config} `config`
 */
const defineBuildConfig = (config: Config = {}): Config => config

export default defineBuildConfig
