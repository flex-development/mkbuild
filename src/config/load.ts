/**
 * @file Config - loadBuildConfig
 * @module mkbuild/config/loadBuildConfig
 */

import type { Config } from '#src/interfaces'
import { cosmiconfig } from 'cosmiconfig'
import es from './loader-es'

/**
 * Searches for a build configuration file.
 *
 * If found, a config object will be returned. The object will contain options
 * from the config file. If not found, an empty object will be returned.
 *
 * @see https://github.com/davidtheclark/cosmiconfig#explorersearch
 *
 * @async
 *
 * @param {string} [location] - Directory to search
 * @return {Config} Build configuration options
 */
const loadBuildConfig = async (location: string): Promise<Config> => {
  /**
   * Module name.
   *
   * @const {string} name
   */
  const name: string = 'build'

  // get search function
  const { search } = cosmiconfig(name, {
    loaders: { '.cts': es, '.js': es, '.mjs': es, '.mts': es, '.ts': es },
    searchPlaces: [
      '#.config.cjs',
      '#.config.cts',
      '#.config.js',
      '#.config.json',
      '#.config.mjs',
      '#.config.mts',
      '#.config.ts'
    ].map(place => place.replace('#', name)),
    stopDir: location
  })

  return ((await search(location))?.config as Config | null) ?? {}
}

export default loadBuildConfig
