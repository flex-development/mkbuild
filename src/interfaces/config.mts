/**
 * @file Interfaces - Config
 * @module mkbuild/interfaces/Config
 */

import type { FileSystem, Task } from '@flex-development/mkbuild'

/**
 * Build configuration.
 *
 * @see {@linkcode Task}
 *
 * @extends {Task}
 */
interface Config extends Task {
  /**
   * File system API.
   *
   * @see {@linkcode FileSystem}
   */
  fs?: Partial<FileSystem> | null | undefined

  /**
   * Build tasks.
   *
   * Each build task will be merged with the base task. Top-level lists (arrays
   * and sets) are concatted and deduped. Plain objects are deep merged.
   *
   * > ðŸ‘‰ **Note**: If empty, `null`, or `undefined`, a single build task will
   * > be inferred from the base task.
   *
   * @see {@linkcode Task}
   */
  tasks?: Task[] | null | undefined
}

export type { Config as default }
