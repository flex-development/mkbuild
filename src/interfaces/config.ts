/**
 * @file Interfaces - Config
 * @module mkbuild/interfaces/Config
 */

import type { FileSystemAdapter } from '#src/types'
import type * as esbuild from 'esbuild'
import type Options from './options'
import type Task from './task'

/**
 * Build configuration options.
 *
 * @extends {Options}
 */
interface Config extends Options {
  /**
   * Load build config from config file.
   *
   * @default true
   */
  configfile?: boolean

  /**
   * Build entries.
   *
   * **Note**: If empty or `undefined`, a single build entry will be inferred
   * from the remaining set of configuration options.
   *
   * @default []
   */
  entries?: Partial<Omit<Task, 'write'>>[]

  /**
   * Custom implementations of file system methods.
   *
   * @see {@linkcode FileSystemAdapter}
   * @see https://nodejs.org/api/fs.html
   *
   * @default fsa
   */
  fs?: FileSystemAdapter

  /**
   * Serve files.
   *
   * @see {@linkcode esbuild.ServeOptions}
   * @see https://esbuild.github.io/api/#serve
   *
   * @default false
   */
  serve?: esbuild.ServeOptions | boolean

  /**
   * Watch files.
   *
   * @default false
   */
  watch?: boolean

  /**
   * Write build results.
   *
   * @default false
   */
  write?: boolean
}

export type { Config as default }
