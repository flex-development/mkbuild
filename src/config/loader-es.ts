/**
 * @file Config - esLoader
 * @module mkbuild/config/esLoader
 */

import type { Config } from '#src/interfaces'
import tsconfigPaths from '#src/plugins/tsconfig-paths/plugin'
import { build } from 'esbuild'
import { resolveImports, toDataURL } from 'mlly'
import { RESOLVE_EXTENSIONS } from './constants'

/**
 * Loads an [ESM][1] (`*.js`, `*.mjs`) or TypeScript (`*.cts`, `*.mts`, `*.ts`)
 * build config file.
 *
 * [1]: https://nodejs.org/api/esm.html
 *
 * @see https://github.com/davidtheclark/cosmiconfig#loaders
 *
 * @async
 *
 * @param {string} path - Config file path
 * @param {string} content - Config file content
 * @return {Promise<Config>} Build configuration options
 */
const esLoader = async (path: string, content: string): Promise<Config> => {
  const {
    outputFiles: [output]
  } = await build({
    entryPoints: [path],
    format: 'esm',
    loader: {
      '.cts': 'ts',
      '.js': 'js',
      '.mjs': 'js',
      '.mts': 'ts',
      '.ts': 'ts'
    },
    metafile: true,
    outdir: '<stdout>',
    plugins: [tsconfigPaths()],
    write: false
  })

  // resolve imports in content to evaluate module
  content = await resolveImports(output!.text, {
    extensions: RESOLVE_EXTENSIONS,
    url: path
  })

  return ((await import(toDataURL(content))) as { default: Config }).default
}

export default esLoader
