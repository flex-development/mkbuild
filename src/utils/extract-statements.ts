/**
 * @file Utilities - extractStatements
 * @module mkbuild/utils/extractStatements
 */

import type { Format } from 'esbuild'
import { findDynamicImports, findExports, findStaticImports } from 'mlly'
import { REQUIRE_REGEX } from 'src/config/constants'
import type { Statement } from 'src/interfaces'

/**
 * Returns an array containing `export`, `import`, and/or `require` statements
 * in `contents`.
 *
 * If `format` is `esm`, the returned array will contain `export` statements, as
 * well as dynamic and static `import` statements. Otherwise, the returned array
 * will contain `require` and dynamic `import` statements.
 *
 * @param {string} [code=''] - Code to extract statements from
 * @param {Format} [format='esm'] - `code` format
 * @return {Statement[]} `import`, `export`, and `require` statements
 */
const extractStatements = (
  code: string = '',
  format: Format = 'esm'
): Statement[] => {
  if (!code) return []

  /**
   * ESM format check.
   *
   * @const {boolean} esm
   */
  const esm: boolean = format === 'esm'

  /**
   * Import, export, and require statements.
   *
   * @const {Statement[]} statements
   */
  const statements: Statement[] = [
    ...findDynamicImports(code),
    ...(esm ? findExports(code) : []),
    ...(esm ? findStaticImports(code) : [])
  ].map(obj => ({
    code: obj.code,
    end: obj.end,
    specifier: obj.type === 'dynamic' ? obj.expression : obj.specifier,
    start: obj.start,
    type: obj.type
  }))

  // push require statements
  if (!esm) {
    for (const match of code.matchAll(REQUIRE_REGEX)) {
      const { 0: code, 1: specifier, index } = match

      statements.push({
        code: code!,
        end: index! + code!.length,
        specifier: specifier!,
        start: index!,
        type: /require\.resolve\(["']\w+["']\)/.test(code!)
          ? 'require.resolve'
          : 'require'
      })
    }
  }

  return statements
}

export default extractStatements
