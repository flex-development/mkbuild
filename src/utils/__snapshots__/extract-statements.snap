// Vitest Snapshot v1

exports[`unit:utils/extractStatements > export > should return declaration export statement array 1`] = `
[
  {
    "code": "export const foo",
    "end": 16,
    "specifier": undefined,
    "start": 0,
    "type": "declaration",
  },
]
`;

exports[`unit:utils/extractStatements > export > should return declare export statement array 1`] = `
[
  {
    "code": "export declare const foo",
    "end": 24,
    "specifier": undefined,
    "start": 0,
    "type": "declaration",
  },
]
`;

exports[`unit:utils/extractStatements > export > should return default export statement array 1`] = `
[
  {
    "code": "export default ",
    "end": 15,
    "specifier": undefined,
    "start": 0,
    "type": "default",
  },
]
`;

exports[`unit:utils/extractStatements > export > should return named export statement array 1`] = `
[
  {
    "code": "export { bar, baz }",
    "end": 19,
    "specifier": undefined,
    "start": 0,
    "type": "named",
  },
]
`;

exports[`unit:utils/extractStatements > export > should return star export statement array 1`] = `
[
  {
    "code": "export * from '#src'",
    "end": 20,
    "specifier": "#src",
    "start": 0,
    "type": "star",
  },
]
`;

exports[`unit:utils/extractStatements > export > should return type export statement array 1`] = `
[
  {
    "code": "export type OrNull",
    "end": 18,
    "specifier": undefined,
    "start": 0,
    "type": "declaration",
  },
]
`;

exports[`unit:utils/extractStatements > import > dynamic > should return dynamic import statement array 1`] = `
[
  {
    "code": "import('bar')",
    "end": 31,
    "specifier": "'bar'",
    "start": 18,
    "type": "dynamic",
  },
]
`;

exports[`unit:utils/extractStatements > import > static > should return default import statement array 1`] = `
[
  {
    "code": "import foo from 'baz'",
    "end": 21,
    "specifier": "baz",
    "start": 0,
    "type": "static",
  },
]
`;

exports[`unit:utils/extractStatements > import > static > should return named import statement array 1`] = `
[
  {
    "code": "import { bar } from 'baz'",
    "end": 25,
    "specifier": "baz",
    "start": 0,
    "type": "static",
  },
]
`;

exports[`unit:utils/extractStatements > require > should return require statement array 1`] = `
[
  {
    "code": "require('bar')",
    "end": 14,
    "specifier": "bar",
    "start": 0,
    "type": "require",
  },
]
`;

exports[`unit:utils/extractStatements > require > should return require.resolve statement array 1`] = `
[
  {
    "code": "require.resolve('baz')",
    "end": 22,
    "specifier": "baz",
    "start": 0,
    "type": "require.resolve",
  },
]
`;
